File: csharp/src/LongRangeForecasts.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace IceCreamScorer
{
    public class LongRangeForecasts
    {
        private readonly Scorer scorer;

        public LongRangeForecasts(Scorer scorer)
        {
            this.scorer = scorer;
        }
        public Dictionary<IceCream, int> LongRangeForecast(string quarter) {
                
            var result = new Dictionary<IceCream, int>();
            if (quarter == "Q1" || quarter == "Q4") {
                foreach (IceCream flavour in Enum.GetValues(typeof(IceCream))) {
                    result[flavour] = 5;
                }
            }
            foreach (IceCream flavour in Enum.GetValues(typeof(IceCream))) {
                result[flavour] = 0;
            }
            List<DateTime> interestingDates = new List<DateTime>() {
                DateTime.Parse("2023-05-01T14:00:00.000-07:00"),
                DateTime.Parse("2023-05-18T14:00:00.000-07:00"),
                DateTime.Parse("2023-06-05T14:00:00.000-07:00"),
                DateTime.Parse("2023-06-23T14:00:00.000-07:00")
            };
            List<bool> expectedWeather = new List<bool>();
            foreach (DateTime day in interestingDates) {

                scorer.UpdateSelection();
                var daysForward = (day - DateTime.Parse("2023-04-26T14:00:00.000-07:00")).Days;
                bool weather = scorer.LookupWeather(daysForward:daysForward);
                expectedWeather.Add(weather);
            }

            foreach (IceCream flavour in result.Keys) {
                result[flavour] += 10;
                var sunnyHolidays = expectedWeather.Select(x => x).Count();
                if (sunnyHolidays > 2 && flavour == IceCream.Vanilla) {
                    result[flavour] +=  5;
                }
                else if (sunnyHolidays > 1 && expectedWeather[0]) {
                    result[flavour] += 2;
                }
            }
            return result;
        }
    }
}

File: python/src/long_range_forecasts.py
import scorer
from scorer import IceCream
import datetime

def long_range_forecast(quarter: str):
    def update_selection_and_lookup_weather(days_forward):
        scorer.update_selection()
        weather = scorer.lookup_weather(days_forward=days_forward)
        return weather

    return long_range_forecast_for_weather(quarter, update_selection_and_lookup_weather)


def long_range_forecast_for_weather(quarter, update_selection_and_lookup_weather):
    result = {}
    if quarter == "Q1" or quarter == "Q4":
        for flavour in IceCream:
            result[flavour] = 5
    for flavour in IceCream:
        result[flavour] = 0
    interesting_dates = [
        datetime.date.fromisoformat("2023-05-01"),
        datetime.date.fromisoformat("2023-05-18"),
        datetime.date.fromisoformat("2023-06-05"),
        datetime.date.fromisoformat("2023-06-23"),
    ]
    expected_weather = []
    for day in interesting_dates:
        days_forward = (day - datetime.date.fromisoformat("2023-04-26")).days
        weather = update_selection_and_lookup_weather(days_forward)
        expected_weather.append(weather)
    for flavour in IceCream:
        result[flavour] += 10
        sunny_holidays = len([True for d in expected_weather if expected_weather])
        if sunny_holidays > 2 and flavour == IceCream.Vanilla:
            result[flavour] += 5
        elif sunny_holidays > 1 and expected_weather[1]:
            result[flavour] += 2
    return result


